provider/docker.go

```diff
@@ -82,56 +124,96 @@ func (provider *Docker) Provide(configurationChan chan<- types.ConfigMessage, po
 			ctx := context.Background()
 			version, err := dockerClient.ServerVersion(ctx)
 			log.Debugf("Docker connection established with docker %s (API %s)", version.Version, version.APIVersion)
-			containers, err := listContainers(ctx, dockerClient)
-			if err != nil {
-				log.Errorf("Failed to list containers for docker, error %s", err)
-				return err
+			var dockerDataList []DockerData
+			if provider.SwarmMode {
+				dockerDataList, err = listServices(ctx, dockerClient)
+				if err != nil {
+					log.Errorf("Failed to list services for docker swarm mode, error %s", err)
+					return err
+				}
+			} else {
+				dockerDataList, err = listContainers(ctx, dockerClient)
+				if err != nil {
+					log.Errorf("Failed to list containers for docker, error %s", err)
+					return err
+				}
 			}
-			configuration := provider.loadDockerConfig(containers)
+
+			configuration := provider.loadDockerConfig(dockerDataList)
 			configurationChan <- types.ConfigMessage{
 				ProviderName:  "docker",
 				Configuration: configuration,
 			}
 			if provider.Watch {
 				ctx, cancel := context.WithCancel(ctx)
-				pool.Go(func(stop chan bool) {
-					for {
-						select {
-						case <-stop:
+				if provider.SwarmMode {
```

Comment:
I wonder how we could make it better. The complexity of this method is getting bigger and bigger :sweat: